waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,468 @@
+
+      let res = catchable.get()
+      let response = res.valueOr:
+        return err("archive error: " & $error)
+
+      let elements = collect(newSeq):
+        for (hash, msg) in response.hashes.zip(response.messages):
+          (hash, msg.timestamp)
+
+      let cursor =
+        if response.cursor.isNone():
+          none(WakuMessageHash)
+        else:
+          some(response.cursor.get().hash)
+
+      return ok((elements, cursor))
+  )
+
+proc initTransferHandler(
+    self: WakuSync, wakuArchive: WakuArchive, wakuStoreClient: WakuStoreClient
+) =
+  if wakuArchive.isNil() or wakuStoreClient.isNil():
+    return
+
+  self.transferCallBack = some(
+    proc(
+        hashes: seq[WakuMessageHash], peerId: PeerId
+    ): Future[Result[void, string]] {.async: (raises: []), closure.} =
+      var query = StoreQueryRequest()
+      query.includeData = true
+      query.messageHashes = hashes
+      query.paginationLimit = some(uint64(100))
+
+      while true:
+        let catchable = catch:
+          await wakuStoreClient.query(query, peerId)
+
+        if catchable.isErr():
+          return err("store client error: " & catchable.error.msg)
+
+        let res = catchable.get()
+        let response = res.valueOr:
+          return err("store client error: " & $error)
+
+        query.paginationCursor = response.paginationCursor
+
+        for kv in response.messages:
+          let handleRes = catch:
+            await wakuArchive.handleMessage(kv.pubsubTopic.get(), kv.message.get())
+
+          if handleRes.isErr():
+            error "message transfer failed", error = handleRes.error.msg
+            # Messages can be synced next time since they are not added to storage yet.
+            continue
+
+          self.ingessMessage(kv.pubsubTopic.get(), kv.message.get())
+
+        if query.paginationCursor.isNone():
+          break
+
+      return ok()
+  )
+
+proc initFillStorage(
+    self: WakuSync, wakuArchive: WakuArchive
+): Future[Result[void, string]] {.async.} =
+  if wakuArchive.isNil():
+    return ok()
+
+  let endTime = getNowInNanosecondTime()
+  let starTime = endTime - self.syncInterval.nanos
+
+  var query = ArchiveQuery(
+    includeData: true,
+    cursor: none(ArchiveCursor),
+    startTime: some(starTime),
+    endTime: some(endTime),
+    pageSize: 100,
+  )
+
+  while true:
+    let response = (await wakuArchive.findMessages(query)).valueOr:
+      return err($error)
+
+    for (topic, msg) in response.topics.zip(response.messages):
+      self.ingessMessage(topic, msg)
+
+    if response.cursor.isNone():
+      break
+
+    query.cursor = response.cursor
+
+  return ok()
+
+proc new*(
+    T: type WakuSync,
+    peerManager: PeerManager,
+    maxFrameSize: int = DefaultMaxFrameSize,
+    syncInterval: timer.Duration = DefaultSyncInterval,
+    relayJitter: Duration = DefaultGossipSubJitter,
+    pruning: bool,
+    wakuArchive: WakuArchive,
+    wakuStoreClient: WakuStoreClient,
```

Comment:
Doesn't this create the kind of interdependency that you were trying to avoid? Even if the `WakuSync` object is not composed with `wakuArchive` and `wakuStoreClient`, it now does need to be aware of these during construction. In my mind Store-Sync is a cleaner abstraction layer that understands the interrelationship between these components without forcing interdependencies into any of these. One way in which this could be achieved without defining a `WakuStoreSync` protocol, would be to simply define the handlers in a separate module and having waku_node install the handlers rather than `WakuSync` doing so during construction.

Reply:
> Doesn't this create the kind of interdependency that you were trying to avoid?

Yes it's meh, but it always been like that, status-quo didn't change. The only benefit would be a cleaner `wakunode` and maybe an easier refactor in the future?

> Even if the WakuSync object is not composed with wakuArchive and wakuStoreClient, it now needs to be aware of these during construction.

Until we have channels or similar, no matter at which layer of abstraction we "construct" it this is always going to be the case.

> In my mind Store-Sync is a cleaner abstraction layer that understands the interrelationship between these components without forcing interdependencies into any of the subcomponents. One way in which this could be achieved without defining a WakuStoreSync protocol, would be to simply define the handlers in a separate module and having waku_node install the handlers rather than WakuSync doing so during construction.

I really don't understand, you should build it and show me. I would love for a better way to exist.